api/v2/serializers/post/token_update.py

```diff
@@ -23,30 +24,61 @@ def validate(self, data):
         IF identity is found validation will:
         - Ensure that user/token produces a valid driver
         """
-        validated_data = data
-        self.validate_token_with_driver(data['provider'], data['username'], data['project_name'], data['token'])
-        return validated_data
+        logger.info("TokenUpdateSerializer:validate called")
+        # create will fail if it is an invalid token, but we need to create an identity first
+        # Although there is no apparent need for this function - it does nothing
+        # It should be here as it is part of the framework.
+        return data
 
-    def create(self, validated_data):
-        identity = self._find_identity_match(validated_data['provider'], validated_data['username'], validated_data['project_name'])
+    def create(self, data):
+        logger.info("TokenUpdateSerializer::create called")
+        identity = self._find_identity_match(data['provider'], data['username'], data.get('project_name'))
         if not identity:
-            identity = self._create_identity(validated_data['provider'], validated_data['username'], validated_data['project_name'], validated_data['token'])
-            return identity
+            identity = self._create_identity(data['provider'],
+                                             data['username'],
+                                             data.get('project_name'),
+                                             data.get('token'))
+        logger.info("cred -> ex_force_auth_token = %s" % (data.get('token')))
+        identity.update_credential(identity, 'key', data['username'], replace=True)
+        identity.update_credential(identity, 'ex_force_auth_token', data.get('token'), replace=True)
+        identity.update_credential(identity,
+                                   'ex_force_auth_url',
+                                   settings.AUTHENTICATION['KEYSTONE_SERVER'],
+                                   replace=True)
 
-        token_cred = identity.credential_set.filter(key='ex_force_auth_token').first()
-        if token_cred:
-            token_cred.value = validated_data['token']
-            token_cred.save()
-        else:
-            identity.credential_set.create(key='ex_force_auth_token', value=validated_data['token'])
+        auth_url = settings.AUTHENTICATION['KEYSTONE_SERVER'] + '/v3'
+        response = requests.get(auth_url + '/auth/tokens',
+                                headers={'x-auth-token': data['token'], 'x-subject-token': data['token']})
+        try:
+            catalog = json.loads(response.text)
+            project = catalog['token']['project']
+        except KeyError:
+            raise serializers.ValidationError("Invalid token passed")
+        endpoint_catalog = catalog['token']['catalog']
+        compute = None
+        for l in endpoint_catalog:
```

Comment:
Please pardon me, I am confused, I feel I need more context to understand what you are asking.  Are you referring to a pipe ('|') or a capital i ('I') or an lowercase L 'l', and on which line?